### PR TITLE
[Core] Get node to connect for driver in global state accessor

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

When a driver starts, it needs to figure out the node manager port, the Raylet socket file path, and the Plasma store socket file path to be able to connect to the local Raylet. (A worker doesn't need this step because the information is included in worker command-line arguments.) This step is required by all language bindings. Currently, the logic is implemented in Python, which makes it hard to be reused by Java and C++ bindings.

In this PR, we move this logic to C++ code (global state accessor) to make it shared by all language bindings.

This PR also removes the mandatory and hardcoded `--node-manager-port 62665` parameter for `ray start --head` when users try the C++ API.

## Related issue number

#16735 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
Can you provide an explanation about what this is and why we need it?

**Comment by kfstorm** [comment]
@wuisawesome Sorry I didn't add a description because it was too late last night. Please see the PR description.

File: cpp/src/ray/util/process_helper.cc [hunk 1]
```diff
@@ -22,12 +25,44 @@ static std::string GetSessionDir(std::string redis_ip, int port, std::string pas
...
+  RAY_CHECK(parts.size() == 2);
+  try {
+    boost::asio::io_service netService;
+    boost::asio::ip::udp::resolver resolver(netService);
+    boost::asio::ip::udp::resolver::query query(boost::asio::ip::udp::v4(), parts[0],
+                                                parts[1]);
+    boost::asio::ip::udp::resolver::iterator endpoints = resolver.resolve(query);
+    boost::asio::ip::udp::endpoint ep = *endpoints;
+    boost::asio::ip::udp::socket socket(netService);
+    socket.connect(ep);
```
Comment (qicosmos) [inline cpp/src/ray/util/process_helper.cc:52]: We'd better use a timeout here, and handle failure.

**Comment by SongGuyang** [comment]
Can you merge master and fix the conflicting files?

File: cpp/src/ray/util/process_helper.cc (hunk 1 above)
Comment (kfstorm) [reply cpp/src/ray/util/process_helper.cc:52]: The original Python code doesn't have a timeout either, and I already handled failure via try-catch.

File: cpp/src/ray/util/process_helper.cc [hunk 2]
```diff
@@ -22,12 +25,44 @@ static std::string GetSessionDir(std::string redis_ip, int port, std::string pas
...
+/// code. See
+/// https://stackoverflow.com/questions/2674314/get-local-ip-address-using-boost-asio.
+///
+/// TODO(kfstorm): Make this function shared code and migrate Python & Java to use this
+/// function.
+///
+/// \param address The IP address and port of any known live service on the network
+/// you care about.
+/// \return The IP address by which the local node can be reached from the address.
+static std::string GetNodeIpAddress(const std::string &address = "8.8.8.8:53") {
```
Comment (SongGuyang) [inline cpp/src/ray/util/process_helper.cc:40]: Can you explain why give a default address "8.8.8.8:53"? Should we make it to be a constant ?

File: cpp/src/ray/util/process_helper.cc [hunk 3]
```diff
@@ -44,28 +79,51 @@ void ProcessHelper::StopRayNode() {
...
   }
 
+  std::string redis_address =
+      redis_ip + ":" + std::to_string(ConfigInternal::Instance().redis_port);
+  std::string node_ip = ConfigInternal::Instance().node_ip_address;
+  if (node_ip.empty()) {
+    if (!ConfigInternal::Instance().redis_ip.empty()) {
+      node_ip = GetNodeIpAddress(redis_address);
+    } else {
+      node_ip = GetNodeIpAddress();
```
Comment (SongGuyang) [inline cpp/src/ray/util/process_helper.cc:99]: `GetNodeIpAddress` will be called twice when `redis_ip == "127.0.0.1"` and `node_ip.empty()`?

**Review by SongGuyang — COMMENTED** [review]
Overall good for me. Someone need to review the GCS part.

File: cpp/src/ray/util/process_helper.cc (hunk 3 above)
Comment (kfstorm) [reply cpp/src/ray/util/process_helper.cc:99]: Yes. The logic is the same in Python. See https://github.com/ray-project/ray/blob/bf3e3225b62e813c04982ba72b9ff8d41c62f6e8/python/ray/_private/services.py#L394 https://github.com/ray-project/ray/blob/bf3e3225b62e813c04982ba72b9ff8d41c62f6e8/python/ray/_private/services.py#L429 https://github.com/ray-project/ray/blob/bf3e3225b62e813c04982ba72b9ff8d41c62f6e8/python/ray/node.py#L105

File: src/ray/gcs/gcs_client/global_state_accessor.cc [hunk 4]
```diff
@@ -277,5 +279,94 @@ std::string GlobalStateAccessor::GetSystemConfig() {
...
+            << "You might need to provide --node-ip-address to specify the IP "
+            << "address that the head should use when sending to this node.";
+        status = Status::NotFound(oss.str());
+      } else {
+        *node_to_connect = nodes[relevant_client_index].SerializeAsString();
+        return Status::OK();
+      }
+    }
+
+    if (!status.IsNotFound()) {
```
Comment (WangTaoTheTonic) [inline src/ray/gcs/gcs_client/global_state_accessor.cc:355]: This check can be removed.

**Review by WangTaoTheTonic — APPROVED** [review]
It doesn't add new function in GCS but only reuse existing `get all nodes`, LGTM with just a tiny comments.
Probably you need to remove the new line added in the end of some files.  :-)

File: cpp/src/ray/util/process_helper.cc (hunk 2 above)
Comment (wuisawesome) [reply cpp/src/ray/util/process_helper.cc:40]: btw 8.8.8.8:53 is Google's primary dns service (which we typically assume is always available). I think it should be available in China (but might be slower?)

File: src/ray/gcs/gcs_client/global_state_accessor.cc (hunk 4 above)
Comment (kfstorm) [reply src/ray/gcs/gcs_client/global_state_accessor.cc:355]: Good catch!

File: cpp/src/ray/util/process_helper.cc (hunk 2 above)
Comment (SongGuyang) [reply cpp/src/ray/util/process_helper.cc:40]: OK, got it!

**Review by SongGuyang — APPROVED** [review]

**Review by wuisawesome — APPROVED** [review]

**Comment by kfstorm** [comment]
`//python/ray/serve:test_deploy` failure on mac should be unrelated. It passed on Linux.

**Pull request merged**